File: Backend/TopSys.TopConWeb.Domain/Services/VisitaService.cs
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using Topsys.TopConWeb.SharedKernel.Common;
using TopSys.TopConWeb.Domain.Entities.Visitas;
using TopSys.TopConWeb.Domain.Interfaces.Repositories;
using TopSys.TopConWeb.Domain.Interfaces.Services;

namespace TopSys.TopConWeb.Domain.Services
{
    public class VisitaService : ServiceBase<Visita>, IVisitaService
    {

        private readonly IVisitaRepository _visitaRepository;

        public VisitaService(IVisitaRepository visitaRepository): base(visitaRepository)
        {
            _visitaRepository = visitaRepository;
        }

        public void AdicionarAnexo(string usuario, Guid id, int usina, int anoVisita, int numeroVisita, string anexo, string nome)
        {
            _visitaRepository.AdicionarAnexo(usuario, id, usina, anoVisita, numeroVisita, anexo, nome);
        }

        public void Adiconar(Visita visita)
        {
            _visitaRepository.Adicionar(visita);
        }

        public void AtualizarDescricaoAnexo(VisitaAnexo anexo)
        {
            _visitaRepository.AtualizarDescricaoAnexo(anexo);
        }

        public ICollection<VisitaAnexo> ListarAnexos(int usina, int anoVisita, int numeroVisita)
        {
            return _visitaRepository.ListarAnexos(usina, anoVisita, numeroVisita);
        }

        public PagedList<Visita> ListarEmOrdemDecrescente(int pagina, int porPagina, Expression<Func<Visita, bool>> filter)
        {
            return _visitaRepository.ListarEmOrdemDecrescente(pagina, porPagina, filter);
        }

        public byte[] ObterAnexo(Guid id)
        {
            return _visitaRepository.ObterAnexo(id);
        }

        public VisitaAnexo ObterVisitaAnexoPorId(Guid id)
        {
            return _visitaRepository.ObterVisitaAnexoPorId(id);
        }

        public Visita ObterPorId(int usina, int ano, int numero, bool tracking = false)
        {
            return _visitaRepository.ObterPorId(usina, ano, numero, tracking);
        }

        public void RemoverAnexo(Guid id)
        {
            _visitaRepository.RemoverAnexo(id);
        }

        public PagedList<VisitaHistorico> ListarHistoricoEmOrdemDecrescente(int pagina, int porPagina, Expression<Func<VisitaHistorico, bool>> filter)
        {
            return _visitaRepository.ListarHistoricoEmOrdemDecrescente(pagina, porPagina, filter);
        }

        public void AdicionarHistorico(VisitaHistorico visitaHistorico)
        {
            _visitaRepository.AdicionarHistorico(visitaHistorico);
        }

    }
}

File: Frontend/src/app/services/visita-service.ts
import { Injectable, Injector } from "@angular/core";
import { BaseService } from "./base.service";
import { Visita } from "app/classes/visita/visita";
import { VisitaAnexo } from "app/classes/visita/visita-anexo";
import { VisitaHistorico } from "app/classes/visita/visita-historico";
import { Lead } from "app/classes/lead/lead";

@Injectable()
export class VisitaService extends BaseService {

    constructor(injector: Injector) {
        super(injector);
    }

    Adicionar(visita: Visita) {
        return this.makePostPrommisse<any>('v1/visita', JSON.stringify(visita));
    }

    Atualizar(visita: Visita) {
        return this.makePatchPrommisse<any>('v1/visita', JSON.stringify(visita));
    }

    Listar(pagina?: number, porPagina?: number, filtro?: string, hideLoading?: boolean) {
        return this.makePagedGetPrommisse<Visita>(`v1/visita?`
            +'pagina='+(pagina ? pagina : 0)
            +'&porPagina='+(porPagina ? porPagina : 0)
            +(filtro ? '&'+filtro : ''),
            hideLoading);
    }

    ObterPorId(usina: number, ano: number, numero: number, hideLoading?: boolean) {
        return this.makeGetPrommisse<Visita>(`v1/visita/usina/${usina}/ano/${ano}/numero/${numero}`, hideLoading);
    }

    obterLeadDeVisita(usina: number, ano: number, numero: number, hideLoading?: boolean) {
        return this.makeGetPrommisse<Lead>(`v1/visita/usina/${usina}/ano/${ano}/numero/${numero}/gerar-lead`, hideLoading);
    }

    adicionarAnexo(usina: number, anoVisita: number, numeroVisita: number, anexo: string, nomeAnexo: string, hideLoading?: boolean) {
        var body = {
            usina: usina,
            ano: anoVisita,
            numero: numeroVisita,
            arquivo: anexo,
            nome: nomeAnexo
        };
        return this.makePostPrommisse<string>('v1/visita/anexo', JSON.stringify(body), hideLoading);
    }

    ObterAnexo(anexoId: string, hideLoading?: boolean){
        return this.makeGetPrommisse<string>('v1/visita/anexo/' + anexoId, hideLoading);
    }

    listarAnexos(usina: number, anoVisita: number, numeroVisita: number, hideLoading?: boolean) {
        return this.makeGetPrommisse<VisitaAnexo[]>('v1/visita/anexo'
        +'/usina/' + usina
        +'/ano/'+ anoVisita
        +'/numero/' + numeroVisita
        , hideLoading);
    }

    atualizarDescricaoAnexo(anexo: VisitaAnexo) {
        return this.makePatchPrommisse<any>('v1/visita/anexo', JSON.stringify(anexo));
    }

    removerAnexo(id: string, hideLoading?: boolean) {
        return this.makeDeletePrommisse<VisitaAnexo>('v1/visita/anexo/' + id, hideLoading);
    }

    adicionarHistorico(historico: VisitaHistorico, hideLoading?: boolean) {
        return this.makePostPrommisse<any>('v1/visita/historico', JSON.stringify(historico), hideLoading);
    }

    ListarHistorico(pagina?: number, porPagina?: number, filtro?: string, hideLoading?: boolean) {
        return this.makePagedGetPrommisse<VisitaHistorico>(`v1/visita/historico?`
            +'pagina='+(pagina ? pagina : 0)
            +'&porPagina='+(porPagina ? porPagina : 0)
            +(filtro ? '&'+filtro : ''),
            hideLoading);
    }
    
}